### ISSUE TITLE
Single user email import

### ISSUE THREAD

**Issue by jomapp** [tutao/tutanota#4837]
As a user I want to **import** my **existing mailbox** from a different provider **into my Tutanota mailbox,** so that I do not lose my old emails and do not have to maintain a second account.

### ✅ Acceptance criteria

- [x] Imported emails shall never be visible in an unencrypted state on the Tutanota servers
- [ ] If an import is interrupted it continues automatically when reconnecting
- [x] All emails existing in the old mailbox should be available in the Tutanota mailbox after the import
- [x] All email data is correctly migrated (correct mapping: sender, recipient, etc.)
- [x] All emails are imported to a separate and newly created folder in Tutanota
- [x] I want to see imported emails in the correct order stored by sent / received timestamp
- [ ] When importing emails I do not lose important information even if the feature is not yet supported in Tutanota
- [x] I can see the progress (amount of emails imported) and if the email import is still running
- [x] I can start the import by storing my IMAP credentials (of my old account) in the settings of my Tutanota account
- [x] My IMAP credentials (of my old account) are stored encrypted within my Tutanota account
- [x] The import is continued once the user logs in
- [x] The import is running as long as it is finished or the user logs out

#### ⛔  Failure modes
- [ ] Not enough (Tutanota) storage available

**Comment by chizutan5** [comment]
Should this not include EML and MBOX import support?

**Comment by rvalitov** [comment]
It's 2023 and I'm surprised a mail service lacks this functionality.... 😞

**Comment by 0x47** [comment]
Has this feature been abandoned? It seems that there has not been any progress in the branch since 4 months. Lack of import functionality keeps me from switching from Protonmail to Tutanota, as I don't want two paid plans in parallel due to email storage space.

**Comment by charlag** [comment]
There was a prototype implementation as a part of a thesis paper but there are some issues that need to be addressed before it can be released.  It is still on our roadmap.

**Comment by 0x47** [comment]
@charlag Can you confirm if there will be a `mbox` or `eml` import function? I need to transfer my mail from ProtonMail to Tutanota and I need to export the mail now, before I cancel my subscription there. Also, is someone actually working on this feature or could this take several years until it is implemented?

**Comment by charlag** [comment]
Email import is currently not available. 
There is a working prototype but we still need to do some changes before it is released.
We are not discussing our ETAs publicly, sorry

**Comment by 0x47** [comment]
Thanks for answering! However, I didn't ask for an ETA, only if there will be (at some point) import-functionality from local files and if there is anyone working on this feature at all. Since you avoided answering, I assume that is not the case.

**Comment by rvalitov** [comment]
@0x47
I'm not in Tutanota team and I've been desperately looking for this feature as a client after buying Premium. After searching I found out that this feature has been announced over 5 years ago with many public ETAs and official release dates that have been overdue. Sorry to say that there's no chance this feature will appear any time soon. Some people say it will never appear because of the Tutanota architecture and the dev team need to rewrite all the code to make it work. I didn't dig deep into the code but I tend to believe it, otherwise such important functionality would appear many years ago.

**Comment by charlag** [comment]
> Thanks for answering! However, I didn't ask for an ETA, only if there will be (at some point) import-functionality from local files and if there is anyone working on this feature at all. Since you avoided answering, I assume that is not the case.

the prototype was concentrating on mass-import via IMAP: https://github.com/tutao/tutanota/pull/5582

but adding single-file shouldn't be a problem.



> @0x47 I'm not in Tutanota team and I've been desperately looking for this feature as a client after buying Premium. After searching I found out that this feature has been announced over 5 years ago with many public ETAs and official release dates that have been overdue. Sorry to say that there's no chance this feature will appear any time soon. Some people say it will never appear because of the Tutanota architecture and the dev team need to rewrite all the code to make it work. I didn't dig deep into the code but I tend to believe it, otherwise such important functionality would appear many years ago.

Please, stop using "this issue was opened X years ago" as a reference or estimate. Many things were planned for the far  future and many more things changed since. This is pointless and misleading.

**Comment by 0x47** [comment]
Unfortunately I already bought 1 year of the Revolutionary plan for two users (Family option, luckily I waited with the third) and now I have two completely useless accounts. I can't move my email from Protonmail to Tutanota and I cannot keep it in Protonmail either because there is a space limit in the free plan.

Well, there is the saying "fool me once, ...". You got me with the unlimited custom domain aliases. Maybe I will check back in some years.

**Comment by napolitano** [comment]
I have been eagerly waiting for years for the Tuta team to finally enable mail import - preferably only locally via the mail client, if possible also via the browser. The same goes for export. Both features are essential to me - one, to enable a reasonable onboarding process, and two, to avoid falling into the lock-in trap. For years, we've been told that it's coming, then that you're developing it, then that it's not so simple - but one thing is missing: proper communication and a transparent timeline.

Frankly, I cannot judge how complex it might be to, for example, import a Thunderbird profile, import a backup in MBOX format, or import a directory with EML files... Maybe it's as complex as nuclear physics. Maybe you prioritize other things - like migrating to a different domain or something similar - but the fact remains that the lack of usable import and export functions has been a dealbreaker for me for years. It's also a fact that I wish you would finally step up and implement it - or at least communicate properly.

**Comment by MEmsho** [comment]
Napolitano has probably said it all... When searching for a solution - don't forget that privacy oriented users  (AKA your loyal customers despite the current limited experience) are already used to being limited in a digital world. We don't mind jumping through a few hoops, even if that means backing up all of our mails to a temporary local solution (like thunderbird, local file container or even separate local solution created by you) and then the importing them through the email client (local) software. Heck...by this stage I'm even willing to import them one by one :-)
I would love it if you would consider this a 2 step process - first a temporary solution for more tech savvy users and then the final one for the masses. That way you could also limit a data overflow to your servers when a simple plug-and-play solution is finally available. To me it seems that this is 100% the main reason why Tuta is not more widely used....People have their entire lives recorded in (mostly) Gmail and are not willing to let that go.
And don't forget, that despite all the praises and support from us - loyal customers despite the limited experience - all of the above commentaries are absolutely in place. This has been announced and postponed too many times...If you don't know how to do it or you have limited financial resources, ask for help in this (or any other opensource) community. Please stop dragging this any further....

**Comment by bemgca** [comment]
I notice that the last two comments here have not received an answer from anyone at Tuta. There is no doubt that importing emails is necessary. There seems to be considerable doubt that Tuta will ever do anything about it. I do not expect Tuta to provide a date that the function will be available. However I do ask that Tuta at least comment to the effect that they are or are not currently working hard on this issue. Has the student mentioned over a year ago left or is he still working on this issue?       Thanks

**Comment by charlag** [comment]
Just like before we are working on some preconditions to be able to release this. The client code is there, the prototype did work but we are still working on addressing things on the backend to make it work properly. @jomapp is still working with us on those things.

We apologize for such a long wait. Thank you for trusting us do get it right.

**Comment by michaelhjulskov** [comment]
Any guess on when the import function will be implemented? 
Will it be on the free version or only the paid version?

**Comment by napolitano** [comment]
I am not affiliated with Tuta, but I was a user who switched to another provider due to the absence of this important feature, ongoing delays, and lack of clear communication.

There has been a clearly articulated need for this feature since at least 2018. Now, in 2024, there is still no usable solution. In my opinion, it remains uncertain whether the feature will be implemented. The concept behind it appears highly academic, somewhat impractical, and possibly over-engineered. Rather than providing a straightforward option in the local client to import MBOX files or similar formats into the existing infrastructure, the current approach seems complex and may not deliver satisfactory results as planned.

A small side note worth mentioning is the concept outlined at the top regarding acceptance criteria, which has not convinced me. It is quite clear that there was an attempt to implement a "secure" IMAP sync with visual feedback. While this can certainly be done, it has numerous drawbacks concerning security, attack vectors, complexity, and the required level of trust. It is evident that there was an aim to create a digital masterpiece rather than a pragmatic solution. A more practical approach would have been to allow direct import of MBOX files or a local mail folder (e.g., from Thunderbird) using the Tuta Desktop Client (or a specially simplified version). This would eliminate the need for fiddling with IMAP credentials, session continuity, etc., and would significantly reduce the complexity.

Personally, I have stopped waiting for it. While Tuta offers some interesting features, there is now sufficient competition that provides comparable services, making it less compelling to wait. Additionally, it has always been possible to use GPG directly on the client to encrypt emails and store them securely with any provider, without needing to store the private key with the mail host.

Of course, this is just my subjective opinion and assessment, and it should not be taken as a general fact.

**Comment by bernardpaulus** [comment]
:wave: 
I am pretty interested about the email import, what is the current status?
It is unclear to me whether https://github.com/tutao/tutanota/pull/5582 was integrated in the master branch as the PR is marked as closed, not merged.

**Comment by h3xbytes** [comment]
Bumping this issue, as I would like to see this implemented.

**Comment by jomapp** [comment]
Thank you for all the discussions and feedback on this issue! 

We did release a first (preview) version **`v262.250122.0`** of mail import, which allows importing  of `.eml` and `.mbox` files, last week. However, importing mails is currently only available in our desktop client and with a Legend or Unlimited plan at the moment. Mail import will be release to ALL paid plans in the future though.

Happy importing :)

**Comment by lehlud** [comment]
Hi @jomapp, it's been a year now since the first preview version got released for users on the legend and/or unlimited plans. When exactly do you plan on making this feature available to the other paid plans?
Thanks!

**Comment by u004a** [comment]
> Hi [@jomapp](https://github.com/jomapp), it's been a year now since the first preview version got released for users on the legend and/or unlimited plans. When exactly do you plan on making this feature available to the other paid plans? Thanks!

Do you get any info on that?

**Comment by T0mmykn1fe** [comment]
Bit of piss take that in 2026 i have to upgrade from one payment tier to the next just to gain email import functionality 👎🏻 :(
Going from 4 euros to 9 euros puts you in Proton territory, where for that price they offer a lot of bells and whistles. I've moved to your platform specifically for 3 custom email domains and a reduced monthly/yearly costing, but this upgrade for import....bad taste in the mouth, guys
